builtins: implement covar_pop and covar_samp aggregation functions

mneverov opened the pull request:
See #41274

Release note (sql change): implement covar_pop and covar_samp aggregation functions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55707)
<!-- Reviewable:end -->

mneverov commented:
To the reviewer:
I noticed that `corr()` function does not have [benchmarks](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/aggregate_builtins_test.go). Should I add benchmarks for `corr()`, and new `covar` functions and if yes, should I add them for all Float Int combinations as well as with/without null, as it done in [tests](https://github.com/cockroachdb/cockroach/pull/55707/files#diff-93074fa3abac255633fa1f1bd9608cf03a52ce57e752279d9c98c4a8e70653d9R321)?
I also tried to add new functions to the optlogic testdata with the command `make testoptlogic FILES='groupby|distsql_agg' TESTFLAGS="-rewrite" `, but testfiles were not changed.

yuzefovich reviewed (commented):
Thanks for working on this! I have some nits and some suggestions, but it looks pretty good already.

> Reviewed 14 of 14 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-:-MK69hHJATp7xDFLS_vr:bisnckc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1418 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK65tFC5C-tGKQZFo0O:-MK65tFD0-7YQXB1CTNV:b1f8s53) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/logictest/testdata/logic_test/aggregate#L1418)):*
> ```
> query FFFFFFFF
> SELECT covar_pop(y, x)::decimal, covar_pop(int_y, int_x)::decimal, covar_pop(y, int_x)::decimal, covar_pop(int_y, x)::decimal,
> 			 covar_samp(y, x)::decimal, covar_samp(int_y, int_x)::decimal, covar_samp(y, int_x)::decimal, covar_samp(int_y, x)::decimal
> ```

nit: looks like there are tabs here, but I believe we only use spaces for indentation.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1343 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK695xeDcay4IR077_i:-MK695xf-DSVOcp57nFZ:b-5gjxvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b26ada64338937124b555c6134c57c1c1fe64a/pkg/sql/logictest/testdata/logic_test/aggregate#L1343)):*
> ```
> SELECT corr(DISTINCT y, x)::decimal, count(DISTINCT y)  FROM statistics_agg_test
> ----
> 0.9326733179802503 4
> ```

It seems like you're thinking that there is a bug with DISTINCT aggregation, but Postgres returns the same result. What is the issue here?

I think that there might be some confusion of what DISTINCT means for aggregation - it applies only and separately to each aggregate function. In this example, for `corr` only the tuples that are distinct on both `y` and `x` columns are aggregated, but for `count` only the tuples that are distinct on `y` column are aggregated.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK67N4B0AUA9-G3RrdK:-MK67N4B0AUA9-G3RrdL:b-7mpyz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/sem/builtins/aggregate_builtins.go#L181)):*
> ```Go
> 
> 	"covar_pop": makeBuiltin(aggProps(),
> 		makeAggOverload([]*types.T{types.Float, types.Float}, types.Float, newCovarPopAggregate,
> ```

nit: I think it'd be nice to define a helper method that be reused to populate builtin overloads for `corr`, `covar_pop`, and `covar_samp` since all of them have the same properties and same possible input types.

Somewhat relatedly, Postgres also takes in decimals as arguments of these functions, and I think we could easily add the support for that too. What do you think?
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68HDCC5SsHhE3cM1v:-MK68HDCC5SsHhE3cM1w:b-3669hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/sem/builtins/aggregate_builtins_test.go#L321)):*
> ```Go
> // testAggregateFunctionResultDeepCopy is a helper function for testing correlation, covariance,
> // and regression aggregation functions.
> func testAggregateFunctionResultDeepCopy(
> ```

nit: I'd maybe add word "regression" to the function name to make it more explicit.
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 328 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68PalE7EOLo-vvowN:-MK68PalE7EOLo-vvowO:b-3ie8xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/sem/builtins/aggregate_builtins_test.go#L328)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 	t.Run(fmt.Sprintf("%s: float float", funcName), func(t *testing.T) {
> 		testAggregateResultDeepCopy(t, newCorrAggregate, makeFloatTestDatum(10), makeFloatTestDatum(10))
> ```

There are several places where `newCorrAggregate` is used, but I think we intended to use `aggFunc` instead, right?
___
*[pkg/storage/enginepb/mvcc3.pb.go, line 0 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68XWA0bS_TE0HPIQ8:-MK68XWA0bS_TE0HPIQ9:b-m450d3) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/storage/enginepb/mvcc3.pb.go#L0)):*
These changes should not be checked in (we have some issues with formatting that I don't know much about, but they shouldn't be a part of the PR).
___
*[pkg/util/hlc/timestamp.pb.go, line 0 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68gVCDdu0gFrkZhyG:-MK68gVCDdu0gFrkZhyH:bjl9adx) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/util/hlc/timestamp.pb.go#L0)):*
Ditto for removing these changes.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I filed #55776 to track the distinct aggregation bug.

mneverov commented on pkg/sql/logictest/testdata/logic_test/aggregate:
```diff
@@ -1349,86 +1351,125 @@ CREATE TABLE statistics_agg_test (
+  (NULL,  NULL, NULL, NULL, NULL, NULL)
+
+query FFFFFFFFFFFFFFF
+SELECT corr(y, x)::decimal, corr(int_y, int_x)::decimal, corr(y, int_x)::decimal, corr(int_y, x)::decimal, corr(dy, dx),
```
I decided to combine corr, covar_pop & covar_samp queries. Please let me know if it is better to keep them separately.

mneverov commented on pkg/sql/sem/builtins/aggregate_builtins.go:
```diff
@@ -1840,12 +2046,12 @@ func (a *regressionAccumulatorBase) add(y float64, x float64) error {
 		if math.IsNaN(x) || math.IsInf(x, 0) {
-			a.sxx = math.NaN()
-			a.sxy = math.NaN()
+			sxx = math.NaN()
```
fixed the bug i introduced in #55268

mneverov commented on pkg/sql/sem/builtins/aggregate_builtins.go:
```diff
@@ -1866,11 +2072,11 @@ func (a *regressionAccumulatorBase) float64Val(datum tree.Datum) (float64, error
 		return float64(*val), nil
 	default:
-		return 0, fmt.Errorf("invalid type %v", val)
+		return 0, fmt.Errorf("invalid type %T (%v)", val, val)
```
That was printing the value, instead of type

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1418 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK65tFC5C-tGKQZFo0O:-MKAqdTR5IaNCM1QzPtW:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/logictest/testdata/logic_test/aggregate#L1418)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: looks like there are tabs here, but I believe we only use spaces for indentation.
</blockquote></details>

:+1: 
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1343 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK695xeDcay4IR077_i:-MKj0Lwz1leiHYFuFp29:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b26ada64338937124b555c6134c57c1c1fe64a/pkg/sql/logictest/testdata/logic_test/aggregate#L1343)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It seems like you're thinking that there is a bug with DISTINCT aggregation, but Postgres returns the same result. What is the issue here?

I think that there might be some confusion of what DISTINCT means for aggregation - it applies only and separately to each aggregate function. In this example, for `corr` only the tuples that are distinct on both `y` and `x` columns are aggregated, but for `count` only the tuples that are distinct on `y` column are aggregated.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK67N4B0AUA9-G3RrdK:-MKAtYLzCJWOpc_rZXnr:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/sem/builtins/aggregate_builtins.go#L181)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I think it'd be nice to define a helper method that be reused to populate builtin overloads for `corr`, `covar_pop`, and `covar_samp` since all of them have the same properties and same possible input types.

Somewhat relatedly, Postgres also takes in decimals as arguments of these functions, and I think we could easily add the support for that too. What do you think?
</blockquote></details>

:+1: 
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 328 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68PalE7EOLo-vvowN:-MKj0Uk6-WYrzgOwro91:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/sql/sem/builtins/aggregate_builtins_test.go#L328)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

There are several places where `newCorrAggregate` is used, but I think we intended to use `aggFunc` instead, right?
</blockquote></details>

Done.
___
*[pkg/storage/enginepb/mvcc3.pb.go, line  at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68XWA0bS_TE0HPIQ8:-MKj0WklAMmoSImjcfkf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/storage/enginepb/mvcc3.pb.go#L)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

These changes should not be checked in (we have some issues with formatting that I don't know much about, but they shouldn't be a part of the PR).
</blockquote></details>

Done.
___
*[pkg/util/hlc/timestamp.pb.go, line  at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK68gVCDdu0gFrkZhyG:-MKj0XXJE2k3lGbxNa_O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23202e1d1ed3006b5f561a22d9764e3eae860cae/pkg/util/hlc/timestamp.pb.go#L)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Ditto for removing these changes.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks, looks very good! I only have a suggestion to support a combination of decimal as one argument and int/float as another for full postgres compatibility.

> Reviewed 6 of 10 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-:-MKkWFbcCkyNZTBi4QrF:b-5dm9a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1374 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK1PLRTCLMASRPYP6Q0-r3-1374:-MKkK3dx-BczUHQMkUPS:b-8b803q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1668926c5bbcdc6e17215696ab8832f53229ea8b/pkg/sql/logictest/testdata/logic_test/aggregate#L1374)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

I decided to combine corr, covar_pop & covar_samp queries. Please let me know if it is better to keep them separately.
</blockquote></details>

Personally, I'd find it easier if the query was split up because now it has many columns and the result values are very wide too.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1373 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkJrRQCwjwHNmhXfBL:-MKkJrRQCwjwHNmhXfBM:bswya7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/logictest/testdata/logic_test/aggregate#L1373)):*
> ```
>   (NULL,  NULL, NULL, NULL, NULL, NULL)
> 
> query FFFFFFFFFFFFFFF
> ```

I think we need to use `F` type only for columns where the function returns `FLOAT`, right? We have a few that return `DECIMAL`, and those should have the exact match using `R` type.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1375 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkHcAh7U0ciw_BwfbQ:-MKkHcAh7U0ciw_BwfbR:b-jzdf8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/logictest/testdata/logic_test/aggregate#L1375)):*
> ```
> query FFFFFFFFFFFFFFF
> SELECT corr(y, x)::decimal, corr(int_y, int_x)::decimal, corr(y, int_x)::decimal, corr(int_y, x)::decimal, corr(dy, dx),
> 			 covar_pop(y, x)::decimal, covar_pop(int_y, int_x)::decimal, covar_pop(y, int_x)::decimal, covar_pop(int_y, x)::decimal, covar_pop(dy, dx),
> ```

nit: also a few tabs in here.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 2049 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK1PLRTCLMASRPYP6Q5-r3-2049:-MKkIkkY0FDG18_uB2IT:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1668926c5bbcdc6e17215696ab8832f53229ea8b/pkg/sql/sem/builtins/aggregate_builtins.go#L2049)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

fixed the bug i introduced in #55268 
</blockquote></details>

Ack.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 2075 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK1PLRTCLMASRPYP6Q5-r3-2075:-MKkIn-Q-vYOxuo2_1uR:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1668926c5bbcdc6e17215696ab8832f53229ea8b/pkg/sql/sem/builtins/aggregate_builtins.go#L2075)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

That was printing the value, instead of type
</blockquote></details>

Ack.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkLAbc1kfXAb8E7S64:-MKkLAbc1kfXAb8E7S65:ba62gou) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L648)):*
> ```Go
> 		makeAggOverload([]*types.T{types.Int, types.Int}, types.Float, aggregateFunc,
> 			info, tree.VolatilityImmutable),
> 		makeAggOverload([]*types.T{types.Decimal, types.Decimal}, types.Decimal, aggregateFuncDecimal,
> ```

I think it'd also be nice support the combination of `decimal` with `int` or `float` on both sides since Postgres supports it.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1758 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkLNt7BYbWHXTB04rB:-MKkLNt7BYbWHXTB04rC:bo395it) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1758)):*
> ```Go
> }
> 
> // regressionAccumulatorDecimalBase is a base struct for the aggregate functions for statistics for decimal numbers.
> ```

nit: we try to keep the lines with 80 character limit, and that is especially applicable to the comments.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1817 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkUGYU4Enlx9wVUCG4:-MKkUGYU4Enlx9wVUCG5:b96f6ls) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1817)):*
> ```Go
> // Reset implements tree.AggregateFunc interface.
> func (a *regressionAccumulatorDecimalBase) Reset(ctx context.Context) {
> 	a.n.SetInt64(0)
> ```

nit: I wonder whether it'd be more efficient to do something like
```
a.n = decimalZero
...
```
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1871 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkUq_8C1wD3HA6uEje:-MKkUq_8C1wD3HA6uEjf:bp0h16l) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1871)):*
> ```Go
> 
> 		// Overflow check.  We only report an overflow error when finite
> 		// inputs lead to infinite results.  Note also that sxx, syy and Sxy
> ```

nit: double space.
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 548 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkVerp8Mx1Atzkq9y9:-MKkVerp8Mx1Atzkq9yA:bc59jwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins_test.go#L548)):*
> ```Go
> }
> 
> const count = 1000
> ```

nit: I'd rename it to something more specific like `aggregateBuiltinsBenchmarkCount` so that it doesn't unwillingly get used in other places in this package.


<!-- Sent from Reviewable.io -->

mneverov commented:
@yuzefovich thanks for the review! I'll address the comments this week

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1374 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MK1PLRTCLMASRPYP6Q0-r3-1374:-MKkrgSU5GpPhVCMZ9iZ:b-18lxgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1668926c5bbcdc6e17215696ab8832f53229ea8b/pkg/sql/logictest/testdata/logic_test/aggregate#L1374)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Personally, I'd find it easier if the query was split up because now it has many columns and the result values are very wide too.
</blockquote></details>

separated corr and covariance back.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1373 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkJrRQCwjwHNmhXfBL:-MKpoQOY89Kibfj1NDCQ:b-9u3my6) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/logictest/testdata/logic_test/aggregate#L1373)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think we need to use `F` type only for columns where the function returns `FLOAT`, right? We have a few that return `DECIMAL`, and those should have the exact match using `R` type.
</blockquote></details>

I agree, but the problem is with distributed queries and because they may be sorted differently the query results may vary. All local tests are passing, but fakedist tests fail because of this.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1758 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkLNt7BYbWHXTB04rB:-MKp5lUF4JDV5gsR8nkD:be4jyq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1758)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: we try to keep the lines with 80 character limit, and that is especially applicable to the comments.
</blockquote></details>

got it, thx
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1817 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkUGYU4Enlx9wVUCG4:-MKkmEWc3rWBNKsyQaeV:bxq0ugx) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1817)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I wonder whether it'd be more efficient to do something like
```
a.n = decimalZero
...
```
</blockquote></details>

yes, you're right.
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 548 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkVerp8Mx1Atzkq9y9:-MKkgavq10pfKsTXd4pg:bbvy489) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins_test.go#L548)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I'd rename it to something more specific like `aggregateBuiltinsBenchmarkCount` so that it doesn't unwillingly get used in other places in this package.
</blockquote></details>

done, thx


<!-- Sent from Reviewable.io -->

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkLAbc1kfXAb8E7S64:-MKq2F1o3gNOihoOavuu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L648)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think it'd also be nice support the combination of `decimal` with `int` or `float` on both sides since Postgres supports it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks! A few more comments from me.

> Reviewed 8 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-:-ML9_02-3EMcQS2T94O3:bqr479x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1373 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkJrRQCwjwHNmhXfBL:-ML9SY4K2Ea2VY9XOSZw:b76jt90) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/logictest/testdata/logic_test/aggregate#L1373)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

I agree, but the problem is with distributed queries and because they may be sorted differently the query results may vary. All local tests are passing, but fakedist tests fail because of this.
</blockquote></details>

Hm, ok, I see. I think the problem is that depending on the order of summation we might get more or less digits in the result, but we could cast to a decimal with specified precision and scale, e.g.
```
query RRRR
SELECT corr(y, dx)::DECIMAL(24, 24), corr(int_y, dx)::DECIMAL(24, 24), corr(dy, int_x)::DECIMAL(24, 24), corr(dy, x)::DECIMAL(24, 24)
FROM statistics_agg_test
----
0.933007822647968026281734  0.933007822647968026281734  0.933007822647968026281734  0.933007822647968026281734
```
I've stressed it for a bit with `fakedist` config, and it works ok. What do you think?

My concern is that the computation with decimals must be precise and by using `F` type we're applying float-type specific comparison incorrectly out of convenience.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkLAbc1kfXAb8E7S64:-ML9THne894h46OylF9r:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L648)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1817 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkUGYU4Enlx9wVUCG4:-ML9Yx1q8PpMoDiS4pFq:b-q40g5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1817)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

yes, you're right.
</blockquote></details>

I think this could be simplified even further:
```
diff --git a/pkg/sql/sem/builtins/aggregate_builtins.go b/pkg/sql/sem/builtins/aggregate_builtins.go
index ca746eec3c..3d907458f2 100644
--- a/pkg/sql/sem/builtins/aggregate_builtins.go
+++ b/pkg/sql/sem/builtins/aggregate_builtins.go
@@ -1829,22 +1829,10 @@ func (a *regressionAccumulatorDecimalBase) Add(
 
 // Reset implements tree.AggregateFunc interface.
 func (a *regressionAccumulatorDecimalBase) Reset(ctx context.Context) {
-       a.n = *decimalZero
-       a.sx = *decimalZero
-       a.sxx = *decimalZero
-       a.sy = *decimalZero
-       a.syy = *decimalZero
-       a.sxy = *decimalZero
-       a.tmpX = *decimalZero
-       a.tmpY = *decimalZero
-       a.scale = *decimalZero
-       a.tmp = *decimalZero
-       a.tmpN = *decimalZero
-       a.tmpSx = *decimalZero
-       a.tmpSxx = *decimalZero
-       a.tmpSy = *decimalZero
-       a.tmpSyy = *decimalZero
-       a.tmpSxy = *decimalZero
+       *a = regressionAccumulatorDecimalBase{
+               singleDatumAggregateBase: a.singleDatumAggregateBase,
+               ed:                       a.ed,
+       }
        a.reset(ctx)
 }
```
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1953 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-ML9ZNq81NqC7EJQQBec:-ML9ZNq81NqC7EJQQBed:b4ozysp) ([raw file](https://github.com/cockroachdb/cockroach/blob/764ea3017569a8957346a162f4c93923ac93c395/pkg/sql/sem/builtins/aggregate_builtins.go#L1953)):*
> ```Go
> 	switch val := datum.(type) {
> 	case *tree.DFloat:
> 		dd := &tree.DDecimal{}
> ```

One optimization we could make is for `decimalVal` to take in another temporary `*apd.Decimal` object that we would use to set the value to instead of incurring an allocation on every `decimalVal` call for floats and ints.


<!-- Sent from Reviewable.io -->

mneverov committed:
builtins: implement covar_pop and covar_samp aggregation functions

See #41274

Release note (sql change): implement covar_pop and covar_samp aggregation functions.

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1373 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkJrRQCwjwHNmhXfBL:-MLBrZV0AKBDMH9axAcX:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/logictest/testdata/logic_test/aggregate#L1373)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, ok, I see. I think the problem is that depending on the order of summation we might get more or less digits in the result, but we could cast to a decimal with specified precision and scale, e.g.
```
query RRRR
SELECT corr(y, dx)::DECIMAL(24, 24), corr(int_y, dx)::DECIMAL(24, 24), corr(dy, int_x)::DECIMAL(24, 24), corr(dy, x)::DECIMAL(24, 24)
FROM statistics_agg_test
----
0.933007822647968026281734  0.933007822647968026281734  0.933007822647968026281734  0.933007822647968026281734
```
I've stressed it for a bit with `fakedist` config, and it works ok. What do you think?

My concern is that the computation with decimals must be precise and by using `F` type we're applying float-type specific comparison incorrectly out of convenience.
</blockquote></details>

Fixed
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1817 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkUGYU4Enlx9wVUCG4:-MLBrPs7-_ggqYLPa9xa:b-5pomlu) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1817)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this could be simplified even further:
```
diff --git a/pkg/sql/sem/builtins/aggregate_builtins.go b/pkg/sql/sem/builtins/aggregate_builtins.go
index ca746eec3c..3d907458f2 100644
--- a/pkg/sql/sem/builtins/aggregate_builtins.go
+++ b/pkg/sql/sem/builtins/aggregate_builtins.go
@@ -1829,22 +1829,10 @@ func (a *regressionAccumulatorDecimalBase) Add(
 
 // Reset implements tree.AggregateFunc interface.
 func (a *regressionAccumulatorDecimalBase) Reset(ctx context.Context) {
-       a.n = *decimalZero
-       a.sx = *decimalZero
-       a.sxx = *decimalZero
-       a.sy = *decimalZero
-       a.syy = *decimalZero
-       a.sxy = *decimalZero
-       a.tmpX = *decimalZero
-       a.tmpY = *decimalZero
-       a.scale = *decimalZero
-       a.tmp = *decimalZero
-       a.tmpN = *decimalZero
-       a.tmpSx = *decimalZero
-       a.tmpSxx = *decimalZero
-       a.tmpSy = *decimalZero
-       a.tmpSyy = *decimalZero
-       a.tmpSxy = *decimalZero
+       *a = regressionAccumulatorDecimalBase{
+               singleDatumAggregateBase: a.singleDatumAggregateBase,
+               ed:                       a.ed,
+       }
        a.reset(ctx)
 }
```
</blockquote></details>

Not sure if `a.reset(ctx)` is necessary in this case.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1953 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-ML9ZNq81NqC7EJQQBec:-ML9gjYv4oRtK5jOXKlH:bnoyiyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/764ea3017569a8957346a162f4c93923ac93c395/pkg/sql/sem/builtins/aggregate_builtins.go#L1953)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

One optimization we could make is for `decimalVal` to take in another temporary `*apd.Decimal` object that we would use to set the value to instead of incurring an allocation on every `decimalVal` call for floats and ints.
</blockquote></details>

Neat! Didn't know how to  deal with this.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for your contribution!

bors r=yuzefovich

> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-:-MLER9Yq8JxBVTNF37W-:b-vpwq4e)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1817 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55707#-MKkUGYU4Enlx9wVUCG4:-MLEPSLQ9t8FdTb97UxK:b-q4fe02) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526ba7bb6202ee6b3d2810328fe62e1f8718c6/pkg/sql/sem/builtins/aggregate_builtins.go#L1817)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

Not sure if `a.reset(ctx)` is necessary in this case.
</blockquote></details>

I think it is necessary in order to update `singleDatumAggregateBase` which shrinks the memory usage.


<!-- Sent from Reviewable.io -->
